client/webserver/site/src/js/doc.ts

```diff
@@ -263,6 +272,11 @@ export default class Doc {
     return threeSigFigs.format(v)
   }
 
+  static formatFourSigFigs (v: number) {
+    if (v >= 1000 || Math.round(v) === v) return oneFractionalDigit.format(v)
+    return fourSigFigs.format(v)
+  }
+
   static formatFiveSigFigs (v: number, prec?: number): string {
     if (v >= 10000) return intFormatter.format(Math.round(v))
     else if (v < 1e5) return fullPrecisionFormatter(prec ?? 8 /* rate encoding factor */).format(v)
```

Comment:
I moved `fourSigFigs` func into `doc.ts` (and renamed it to `formatFourSigFigs `) so it is accessible both in `markets.ts` and `charts.ts`, but turns out there already are `formatThreeSigFigs` and `formatFiveSigFigs ` that are pretty similar but use slightly different formatting rules ... which seems somewhat weird to me. Can't we use same formatting for all of these ?

Reply:
You're right. It is weird. We really need consistent guidelines for display of numerical values. As I've pointed out [elsewhere](https://github.com/decred/dcrdex/pull/2281#issuecomment-1498181498), I typically draw a distinction between "display" values and "data" values. For "display" values, I think sig figs is appropriate, especially for numbers that have decimals (but for large numbers, I wouldn't, for instance, truncate 1,234,567 to 1,230,000). I personally feel like nobody ever needs more than 3 significant figures for display values, but I would entertain arguments for 4 or 5 sig figs too. For "data" values, for which it might be necessary to validate the number against e.g. an order, the full value should always be displayed, out to the number of decimals derived from `UnitInfo` and/or `RateStep`.

Reply:
One problem though is that Javascript seems to prioritize `minimumSignificantDigits` over `maximumFractionalDigits`. That's why we end up with values like `0.00000001000` for `fourSigFigs.format(0.00000001)`. I guess a question I have is whether we can omit the `minimumSignificantDigits` specifier in all cases.

Reply:
Right, that's exactly why we unfortunately have to do this `(v >= 1000)` else if `(v < 1e5)` business.  The ability to use the *significant* digits and *fraction* digits options together effectively is considered an [experiential option](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat#roundingpriority).  https://github.com/decred/dcrdex/pull/2208#discussion_r1134733483

Generally though I find 4 sig figs for smaller values to be the sweet spot, but as you said whole numbers precision should not be lost, only fractional values.

Reply:
I agree with all of the above,

> For "data" values the full value should always be displayed

we can't address this ^ in general by simply changing how we are using/combining different rounding modifiers (_significant_ digits and _fraction_ digits) because [JS Number](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number#number_encoding) itself can't represent what is actually **arbitrary precision Decimal**.

Maybe in practice we don't need to (if those cases when we don't get precise representation for order rate/qty/... are very rare on DEX, and we can adjust those _significant_ digits and _fraction_ digits modifiers such that they fit our use-cases most closely. Otherwise we need to figure out how to handle Decimals precisely (use some 3rd party lib to handle it for us, for example), but that's .

Reply:
As you point out, this is why we use integer representations of quantities and rates on the backend.  These numbers find also their way to the frontend.  Particularly with the rates, these whole numbers do not exceed 2^53 – 1 so there is no precision loss because the values fit into the IEEE 754 64-bit float's significand.

Only when we start _doing math_ with these on the frontend can we run into trouble.  When we divide the whole number "message rate" by the "rate encoding factor" (1e8), we get the decimal value.  If the float string formatters are used with the appropriate "rate step" or even the full 8 fractional digits of precision, it is rounded so that there is no incorrect or misleading information presented (the order of precision loss is far smaller than the number of digits we retain and the formatters do not simply truncate).  We only need to request the appropriate number of digits for a rate based on the market's rate step.